File: prog_praxis/harvey.py
from itertools import combinations, ifilter, imap, izip

def sums(n):
    return [n * (n + 1) / 2,
            n * (n + 1) * (2 * n + 1) / 6,
            (n * (n + 1) / 2) ** 2]

def prop(sub):
    return all(0 == a - b for (a, b) in
               izip(sums(16), map(lambda x: 2 * x,
                                   [sum(sub), sum(x**2 for x in sub),
                                   sum(x**3 for x in sub)])))

if __name__ == "__main__":
    from pprint import pprint
    s = set(range(1, 17))
    print next(imap(lambda sub: [sub, tuple(s.difference(sub))],
              ifilter(prop, combinations(range(1, 17), 8))))

File: prog_praxis/harvey.hs
import Data.List ((\\))

combinations :: [Integer] -> Integer -> [[Integer]]
combinations _ 0 = []
combinations [] _ = []
combinations xs 1 = map (:[]) xs
combinations (x:xs) k = map (x:) (combinations xs (k - 1)) ++ combinations xs k

sums :: Integer -> [Integer]
sums n = [(n * (n + 1)) `div` 2,
     (n * (n + 1) * (2 * n + 1)) `div` 6,
     (head $ sums n) ^ 2]
     --(n * (n + 1) `div` 2) ^2]

prop :: [Integer] -> Bool
prop xs = map (`div` 2) (sums 16) == map sum [xs, map (^2) xs, map (^3) xs]

main :: IO ()
main = mapM_ printPairs filteredCombs where
     printPairs = print . (\ c -> (c, [1..16] \\ c))
     filteredCombs =  filter prop $ combinations [1..16] 8